### DISCUSSION TITLE
Checking in Collections definition

### DISCUSSION THREAD

**Discussion by lukepighetti** [pocketbase/pocketbase#489] [Q&A]
It would be really nice if we could check in our collections definitions into git so we could keep track of these changes and then deploy them when necessary.

**Comment by ganigeorgiev** [comment]
There are 2 aspects to this, depending on how you use PocketBase:

**1. Using a prebuilt PocketBase executable**

You can go to "Settings > Export collections" and download the generated collections json file. It contains everything related to your local collections - general properties, fields schema, api rules, etc. You can commit this file in your repo.

To import back the collection, you can go to "Settings > Import collections" and load the previously exported json file.

_This currently cannot be automated by default because the import may require the user to manually resolve eventual "conflicts" with their local collection changes and therefore using the UI is necessary (we have a simple diff preview and some auto-discrepencies resolvers, like syncing the ids of collections with the same name, fields, etc.)._

---

**2. Using PocketBase as a framework**

When PocketBase is used as a Go framework, you can make use of the `migrate collections` command.
This is expained more in depth in https://pocketbase.io/docs/db-migrations/#collections-migration, but in general, the expected flow for this scenario is the following:

- run the migrate collections command
- build your app, aka. `go build` (the migration will be automatically embedded as part of the final executable)
- push your changes/deploy on prod
- repeat...

_This scenario also suffer from the "eventual import conflict" mentioned in 1, but here the developers have direct access to the db and the [event hooks](https://pocketbase.io/docs/event-hooks/) so they can prevent submitting unwanted env changes._
_There is also an option to "disable/hide" the collections create & edit controls from the Admin UI by enabling the toggle in "Settings > Application > Hide collection create and edit controls"._

---

**Suggestions for better/alternative handling are more than welcomed.**

_I'll move the issue to Q&A discussion._

**Reply by lukepighetti** [reply]
I have an immediate need to make it easy for developers to spin up a local version of PocketBase (executable) with the schema. Any suggestions for how to achieve this?

I am currently exporting the schema manually through the UI but at least having a CLI command to do this would be super useful. 

https://github.com/lukepighetti/pocketbase_flutter_app/blob/main/pb_schema.json

**Reply by ganigeorgiev** [reply]
There is no CLI command to export the schema when you use the prebuilt binary. The `migrate collections` command is available only when you use PocketBase as framework.

The easiest way to load the exported collections json is to do it from the Admin UI - "Settings > Import collections > Load from json".

As stated in 1), this cannot be automated because sometimes there could be conflicts with the existing collections in the receiving environment and the developer will have to resolve them manually (if there are changes, you can review the diff in the UI and adjust them if necessary).

**Reply by pchasco** [reply]
One suggestion that comes to mind is to simply archive your PocketBase directory, including any SQL databases, etc. Or, if you want to get fancier, you could containerize it and host it in a Docker registry. Either of these would enable developers to spin up an instance from a known state.

As far as making changes easy to version control, I would love to see a solution to this that doesn't involve writing migrations in Go. I am personally not a Go developer, and have very little interest in learning Go. 

In my day job, I typically write migrations entirely in SQL that are applied in continuous integration. I can imagine a solution in PocketBase that might apply a sequence of migration scripts against a database that does not involve having to write Go. The challenge there would be that PocketBase abstracts the underlying database structure away from the developer, and providing them a mechanism to alter it without any guardrails could prove problematic.

One possibility might be to use a declarative approach to data migrations that does not involve using SQL scripts directly, but instead JSON objects with migration operations. For example, if a migration required populating some data into a lookup table, and updating some records to use those new lookups:

```json
[
  {
    "op": "insert",
    "collection": "colors",
    "records": [
      {
        "name": "blue",
        "#ref": "blue" // Declarative way to refer back to this row later in the migration. Does not actually update a column
      },
      {
        "name": "green",
        "#ref": "green"
      }
    ]
  },
  {
    "op": "update",
    "collection": "favoriteColors",
    "updates": [
      {
        "where": "active != 0",
        "set": [
          {
            "color": { "#ref": "blue" }
          },
          {
            "otherColumn": 10,
            "yetAnotherColumn": "hello, world!"
          }
        ]
      },
      // A more compact alternative to an update
      {
        "where": "active != 0",
        "columns": ["color", "otherColumn", "yetAnotherColumn"],
        // Notice the value for "color" is an object that the migration can recognize as a command to use a ref
        "values": [{ "#ref": "blue" }, 10, "hello, world!"]
      },
      {
        // An update including JSON field
        "where": "active != 0",
        "columns": ["jsonColumn"],
        "values": [{ "json": { "this": "is", { "a": "json object that will be encoded by the migration" }}}]
      }
    ]
  }
]
```

Obviously, this is a very rough idea at the moment. However, the basic idea is that we could write data migrations that can be applied incrementally, in the correct sequence with our schema changes. and the migrations should not require imperative code.

**Reply by ganigeorgiev** [reply]
@pchasco Data import/export will be handled in https://github.com/pocketbase/pocketbase/issues/48.

Collections schema changes is currently supported either by export/import of JSON from the UI or via a automated migration command when used as framework. Why the current "Export -> Import from JSON" approach is not suitable? In the future we can add a command to generate the JSON and to import from cli, but I expected users to prefer the UI for that when not used as framework.

**Reply by pchasco** [reply]
I was interested in providing a way to perform a complex migration declaratively without needing to write a program in Go. I think the current migrations runner is probably sufficient to perform any needed migrations, just as long as you are comfortable writing a Go program to do the migration.

**Reply by lukepighetti** [reply]
Some feedback: a CLI command to export the collection schema JSON built into the binary would be very useful for me and I don't think it's a breaking change or a heavy lift. Get the same result as clicking the button in the UI.

That would allow me to create a pre-commit check to export it and check it in. Then consumers could import via UI.

A 'stretch goal' for this would be to have `pocketbase serve` have a flag to automatically export this JSON when it's changed in the UI, so the correct schema is always in version control without any extra automations. But that is simply not necessary.

**Reply by lukepighetti** [reply]
Only reason I didn't want to check in the `pb_data` folder was because I didn't want exposed secrets, passwords, emails, etc. Even on a local test environment. Maybe it's a non-issue?

**Reply by pchasco** [reply]
In my opinion, any solution that ultimately requires using the dashboard to deploy to a new environment is an unworkable solution for serious development. What would be ideal is if there were a way to perform migrations as a part of a CI/CD pipeline. I am not personally bothered by the idea that the service might be taken offline or placed in a read-only mode while a deployment/migration is in progress. After all, this product doesn't seem intentioned to be a high-availability solution. There are other BaaS out there that can fill that role. A few seconds of downtime or read-only mode here and there is more than acceptable in my opinion.

**Reply by lukepighetti** [reply]
Even just kicking an error back for the time being that says a migration must be done in the UI is fine. AFAIK there doesn't always have to be a migration. But yes, I agree that we should be able to push migrations via CI

**Reply by ganigeorgiev** [reply]
@lukepighetti, @pchasco

Just to let you know that with [PocketBase v0.9.0](https://github.com/pocketbase/pocketbase/releases/tag/v0.9.0) we now have **automigrations support**, which means that any collections change will result in a `pb_migrations/*` file that could be committed and added in your repo.
We also now support writing custom external migrations with JavaScript which should allow running custom db change scripts, data imports, initializing default app settings, etc. with API similar to the Go one (_more documentation will be available soon_).

**Reply by gBasil** [reply]
@ganigeorgiev Especially when prototyping a new project, a lot of unneeded migrations are created. What's the best way to deal with that? Is there a way to automatically merge them into one?

**Reply by ganigeorgiev** [reply]
@gBasil There is no automatic "squash" command, but you can create a complete collections snapshot using [`migrate collections`](https://pocketbase.io/docs/migrations/#migrate-collections).

Then you can delete the previous unnecessary migration files manually and optionally you can call [`migrate history-sync`](https://pocketbase.io/docs/migrations/#migrate-history-sync) to cleanup the deleted entries from your local `_migrations` history table.

The migration process (both JS and Go) will be documented better with the upcoming v0.17.0 release (_there will be also ambient TS declarations for the JS migrations to assist with the codeintellisense_).
